Added new topic for create custom rest api (#9161)
magento/devdocs: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md

---
group: rest-api
title: Create custom rest api tutorial
contributor_name: Ziffity
contributor_link: https://www.Ziffity.com/
---

## Overview

> REST API in Magento 2 defines a set of functions used by the developers to perform requests and get responses using the HTTP protocol. By default, Magento 2 provides some predefined rest APIs like Product API, Order API, Customer data API with which you can take virtual control of everything happening on the site.

Reviewer: ```suggestion
The REST API in Magento 2 defines a set of functions used by the developers to perform requests and get responses using the HTTP protocol. By default, Magento 2 provides some predefined rest APIs like Product API, Order API, Customer data API, with which you can take virtual control of everything happening on the site.
```

> Even though Magento 2 has these many rest APIs, they are not enough when it comes to using the custom data and their values. To manage your custom data and fields, you need to create custom rest API in Magento 2. This tutorial describes how you can create custom REST API.

Reviewer: ```suggestion
Even though Magento 2 has many REST APIs, they are not enough when it comes to using the custom data and their values. To manage your custom data and fields, you need to create a custom REST API in Magento 2. This tutorial describes how you can create such a custom REST API.
```

## 1. Create a custom module

> Everything starts with a module. `Dev_RestApi` will be used as the namespace:

Reviewer: ```suggestion
`Dev_RestApi` will be used as the namespace:
```

            <module name="Magento_Catalog" />
        </sequence>
    </module>
</config>
```

`app/code/Dev/RestApi/registration.php`:

```php
\Magento\Framework\Component\ComponentRegistrar::register(
    \Magento\Framework\Component\ComponentRegistrar::MODULE,
    'Dev_RestApi',
    __DIR__
);
```

## 2. Create custom ACL entries

`app/code/Dev/RestApi/etc/acl.xml`:

```xml
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:Acl/etc/acl.xsd">
    <acl>
        <resources>
            <resource id="Magento_Backend::admin">
                <resource id="Dev_RestApi::products" title="Dev API - Products"
                          translate="title" sortOrder="110">
                    <resource id="Dev_RestApi::products_get" title="Get product"
                              translate="title" sortOrder="10" />
                    <resource id="Dev_RestApi::products_set_description" title="Set description"
                              translate="title" sortOrder="20" />
                </resource>
            </resource>
        </resources>
    </acl>
</config>
```

> In this tutorial, we have defined acl entries for each endpoints, to have full control over who can access.

Reviewer: ```suggestion
In this tutorial, we have defined ACL entries for each endpoint which give full control over who can access the endpoints.
```

## 3. Define custom endpoints

`app/code/Dev/RestApi/etc/webapi.xml`:

```xml
<?xml version="1.0"?>
<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/webapi.xsd">
    <route url="/V1/rest_dev/getProduct/:id" method="GET">
        <service class="Dev\RestApi\Api\ProductRepositoryInterface" method="getItem" />
        <resources>
            <resource ref="Dev_RestApi::products_get" />
        </resources>
    </route>
    <route url="/V1/rest_dev/setDescription" method="PUT">
        <service class="Dev\RestApi\Api\ProductRepositoryInterface" method="setDescription" />
        <resources>
            <resource ref="Dev_RestApi::products_set_description" />
        </resources>
    </route>
</routes>
```

> *  `route.url` is a URL of the endpoint, the full address would be: `<domain>/rest/<store_code><route.url>`

Reviewer: ```suggestion
*  `route.url` is the URL of the endpoint, the full address would be: `<domain>/rest/<store_code><route.url>`
```

*  `route.method` defines the request method. You should follow the rules of the REST API specification:

| Method | Description |
| --------- | ----------- |
| `GET` | To receive data |
| `POST` | To create new object(s) |
| `PUT` | To update existing object(s) |
| `DELETE` | To delete object(s) |

*  `service` informs about the interface and the method to be called when the endpoint is reached.

*  `resource` sets the ACL resource that is required to have access to the endpoint. If you want to make it public (no authentication), you can use: `<resource ref="anonymous" />`

## 4. Create Interfaces

> In this example, we have created interfaces for request and response

Reviewer: ```suggestion
In this example, we have created the interfaces for request and response
```

    /**
     * @return string
     */
    public function getName();

    /**
     * @return string|null
     */
    public function getDescription();

    /**
     * @param int $id
     * @return $this
     */
    public function setId(int $id);

    /**
     * @param string $sku
     * @return $this
     */
    public function setSku(string $sku);

    /**
     * @param string $name
     * @return $this
     */
    public function setName(string $name);

    /**
     * @param string $description
     * @return $this
     */
    public function setDescription(string $description);
}
```
`app/code/Dev/RestApi/Api/ProductRepositoryInterface.php`:

```php
namespace Dev\RestApi\Api;

interface ProductRepositoryInterface
{
    /**
     * Return a filtered product.
     *
     * @param int $id
     * @return \Dev\RestApi\Api\ResponseItemInterface
     * @throws \Magento\Framework\Exception\NoSuchEntityException
     */
    public function getItem(int $id);

    /**
     * Set descriptions for the products.
     *
     * @param \Dev\RestApi\Api\RequestItemInterface[] $products
     * @return void
     */
    public function setDescription(array $products);
}
```

## 4. Create Models

> The objective of models to create classes that implement our interfaces and process data.

Reviewer: ```suggestion
The objective of models is to create classes that implement the interfaces and process data.
```

> In this example, we have created models for request, response and process data. In model we have defined two methods `getItem` which provide product details of the given product id and `setDescription` which updates the description of the given product item(s).

Reviewer: ```suggestion
In this example, we have created models for request, response and process data. In each model we have defined two methods `getItem`, which provides product details of the given product id, and `setDescription` which updates the description of the given product item(s).
```

Author: Hi @BarnyShergold, I have made changes, Kindly review it

`app/code/Dev/RestApi/Model/Api/RequestItem.php`:

```php
namespace Dev\RestApi\Model\Api;

use Dev\RestApi\Api\RequestItemInterface;
use Magento\Framework\DataObject;

/**
 * Class RequestItem
 */
class RequestItem extends DataObject implements RequestItemInterface
{
    /**
     * @return int
     */
    public function getId()
    {
        return $this->_getData(self::DATA_ID);
    }

    /**
     * @return string
     */

>     public function getDescription()

Reviewer: ALL methods should have the return type on the prototype line
